### PR TITLE
feat(snql) Validate SnQL subscriptions

### PR DESCRIPTION
**Pull request by evanh**
Add logic to validate SnQL queries when they're created. When a Delegate
subscription is created, validate both the legacy and SnQL queries.

### CODE REVIEW

**Comment by evanh** [comment]
Note that I didn't include the setting to toggle SnQL queries on right now. This PR already felt like a lot.

File: tests/datasets/validation/test_entity_validation.py [hunk 1]
```diff
@@ -1,7 +1,8 @@
 import datetime
```
Comment (evanh) [inline tests/datasets/validation/test_entity_validation.py:1]: Moved into the folder with the other validator tests.

File: snuba/subscriptions/data.py [hunk 2]
```diff
@@ -175,68 +188,98 @@ def to_dict(self) -> Mapping[str, Any]:
...
+    def validate_subscription(
+        self, query: Union[CompositeQuery[Entity], Query]
+    ) -> None:
+        # TODO: Support composite queries with multiple entities.
+        from_clause = query.get_from_clause()
+        if not isinstance(from_clause, Entity):
+            raise InvalidSubscriptionError("Only simple queries are supported")
+        entity = get_entity(from_clause.key)
+
+        SubscriptionAllowedClausesValidator().validate(query)
```
Comment (evanh) [inline snuba/subscriptions/data.py:257]: Validators felt like the right place to put this logic, but I'm open to suggestions.

File: snuba/subscriptions/data.py [hunk 3]
```diff
@@ -175,68 +188,98 @@ def to_dict(self) -> Mapping[str, Any]:
...
+        OneProjectValidator().validate(query)
```
Comment (fpacifici) [inline snuba/subscriptions/data.py:258]: Why do we need this validator? 
The add_conditions method adds a `project_id = a_value` condition. This means we are always structurally certain there will be a project_id condition that will not be too wide. 
If somebody adds a condition on multiple projects, the query will simply return nothing as it does today.

File: snuba/query/validation/validators.py [hunk 4]
```diff
@@ -52,3 +55,69 @@ def validate(self, query: Query, alias: Optional[str] = None) -> None:
...
+                project_id: Optional[int] = None
+                if isinstance(rhs, Literal) and isinstance(rhs.value, int):
+                    project_id = rhs.value
+                elif isinstance(rhs, FunctionCall):
+                    project_ids = set(rhs.parameters)
+                    if len(project_ids) != 1:
+                        raise InvalidQueryException("Must only query one project")
+                    value = project_ids.pop()
+                    assert isinstance(value, Literal) and isinstance(value.value, int)
+                    project_id = value.value
```
Comment (fpacifici) [inline snuba/query/validation/validators.py:79]: You should probably raise InvalidQueryException if the set does not contain a Literal that contains an integer.

File: snuba/query/validation/validators.py [hunk 5]
```diff
@@ -52,3 +55,69 @@ def validate(self, query: Query, alias: Optional[str] = None) -> None:
...
+
+                if matched_id is not None and matched_id != project_id:
+                    raise InvalidQueryException("Must only query one project")
```
Comment (fpacifici) [inline snuba/query/validation/validators.py:82]: There is a subclass of this exception (I think it is InvalidExpressionException) that allows you to pass the invalid expression. So we can troubleshoot failures.

File: snuba/query/validation/validators.py [hunk 6]
```diff
@@ -52,3 +55,69 @@ def validate(self, query: Query, alias: Optional[str] = None) -> None:
...
+class SubscriptionAllowedClausesValidator(QueryValidator):
+    def validate(self, query: Query, alias: Optional[str] = None) -> None:
+        selected = query.get_selected_columns()
+        if len(selected) != 1:
+            raise InvalidQueryException("Can only have one aggregation in the select")
+
+        disallowed = ["groupby", "having", "orderby"]
+        for field in disallowed:
+            if getattr(query, f"get_{field}")():
+                print(field, getattr(query, f"get_{field}")())
```
Comment (fpacifici) [inline snuba/query/validation/validators.py:120]: Please remove the print.

File: snuba/subscriptions/subscription.py [hunk 7]
```diff
@@ -29,8 +34,13 @@ def __init__(self, dataset: Dataset):
...
     def create(self, data: SubscriptionData, timer: Timer) -> SubscriptionIdentifier:
         # We want to test the query out here to make sure it's valid and can run
-        request = data.build_request(self.dataset, datetime.utcnow(), None, timer)
-        parse_and_run_query(self.dataset, request, timer)
+        # If there is a delegate subscription, we need to run both the SnQL and Legacy validator
+        if isinstance(data, DelegateSubscriptionData):
+            self._test_request(data.to_snql(), timer)
+            self._test_request(data.to_legacy(), timer)
+        else:
+            self._test_request(data, timer)
```
[the remark this answers is not in the document]
Comment (evanh) [reply snuba/subscriptions/subscription.py:42]: Yes that's all correct. I'm going to add code to the Sentry side to automatically convert new legacy subscriptions to delegate subscriptions (that's why I opened https://github.com/getsentry/snuba-sdk/pull/61).

File: snuba/subscriptions/data.py (hunk 3 above)
Comment (evanh) [reply snuba/subscriptions/data.py:258]: OK sounds good. I had a note from our discussions saying we should validate this but I'll remove it.

Comment (fpacifici) [reply snuba/subscriptions/data.py:258]: Yeah I remember that discussion. Sorry I forgot we were actually adding the condition every time we would actually run this validator.

Comment (evanh) [reply snuba/subscriptions/data.py:258]: It's removed.

File: snuba/subscriptions/data.py [hunk 8]
```diff
@@ -116,7 +125,11 @@ class LegacySubscriptionData(SubscriptionData):
     aggregations: Sequence[Aggregation]
 
     def build_request(
-        self, dataset: Dataset, timestamp: datetime, offset: Optional[int], timer: Timer
+        self,
+        dataset: Dataset,
+        timestamp: datetime,
+        offset: Optional[int],
+        timer: Timer,
```
[the remark this answers is not in the document]
Comment (evanh) [reply snuba/subscriptions/data.py:132]: It's my VSCode set up, I think I have a bad config somewhere. I haven't been able to track it down. It reformats stuff like this and the linter doesn't always care.

**Review by fpacifici — APPROVED** [review]

**Pull request merged**